### ISSUE TITLE
Memory crash after using output repeatedly on large datasets

### ISSUE THREAD

**Issue by vincenzocaselli** [deeplearning4j/deeplearning4j#4917]
During training, when the overall error goes below a given value (e.g. 0.1), I periodically test results using
MultiLayerNetwork.output()
over around 100 test images, in order to have a custom evaluation of the network training.

However, after a few of such testings iterations the memory starts slowly raising until the JVM crashes without any error message in Java console.
I double checked how I instantiate Java new objects in this phase and cleaned every usage until I now have a static array of INDArray (one for every test image) and a correspondent static array of Strings for the labels: they are prepared once in advance and during training just tested with output() method.
Nevertheless the memory still increase as soon as I start using ouput().

Please note that the number of inputs for training is quite high (over 300,000), the network is a CNN (very similar to the Mnist CNN example included in DL4J examples) and overall it behaves just wonderfully, as soon as I do not start using .output() method.

I experienced this behavior both on Linux CentOS 7.4 64 bits and on Ubuntu 16.04 64 bits. 
The DL4J version is 0.9.1, no CUDA/NVIDIA used. 
RAM is 16 GB + 64 GB swap available. 
Memory options in launch are
-XX:+UseSerialGC -Xmx20G -Dorg.bytedeco.javacpp.maxbytes=16G -Dorg.bytedeco.javacpp.maxphysicalbytes=16G

I could of course run this periodic testing session in a separate JVM instance, so to avoid crash of the training instance, but still am wondering what is causing the memory problem.

Could the output method() be affected by some form of memory leak or is there any cleanup I could do after each usage? (I already tried calling .cleanup() and also .detach() on output, but without success) 

Thank you very much
Vincenzo

(initially posted on the new Eclipse Deeplearning4j forum https://www.eclipse.org/forums/index.php/t/1092711/)

**Comment by vincenzocaselli** [comment]
Ok, I'm trying to adapt to 1.0.0-alpha APIs and have a question.
In 0.9.1 I used to do the following:
```
    	INDArray input = Nd4j.create(new int[]{n, ..
    	INDArray output = Nd4j.create(new int[]{n, .    	
    	//Fill input and output for Training
    	fillInputOutput(input, output);
     	DataSet dataSet = new DataSet(input, output);
        List<DataSet> listDs = dataSet.asList();
        Collections.shuffle(listDs,rng);
        return new ListDataSetIterator(listDs,batchSize);
```
In 1.0.0-alpha there's no ListDataSetIterator class.
How can I adapt the last line with new APIs ?

**Comment by vincenzocaselli** [comment]
Thank you @saudet, I wasn't including the new deeplearning4j-utility-iterators, that's why I didn't find ListDataSetIterator.
In the meantime I tried  
TestDataSetIterator( ... ) from nd4j-api, which has the same constructor signature (I know, this is not a good argument for replacing with success, but I tried anyway..)

Well, I am happy to say that @raver119 hit the point: in about 24 hours of training with heavy usage of output() calls (same code, same training set used previously with 0.9.1 version), now with 1.0.0-alpha version the memory usage is stable: RAM is waving between 13-15 GB and swap sticks to about 10 GB, no more growing!
I will of course re-test it all with ListDataSetIterator instead of TestDataSetIterator, but that should not change the behavior about memory.
So, thank you guys: you're just great!
